iotilecore/iotile/core/hw/virtual/tile_based_device.py

```diff
@@ -55,3 +69,207 @@ def stop(self):
+        """Return the hardware identification string."""
+        hardware_string = self.config.get('hardware_string', 'no_version')
+
+        if not isinstance(hardware_string, bytes):
+            hardware_string = hardware_string.encode('utf-8')
+
+        if len(hardware_string) < 10:
+            hardware_string += b'\0'*(10 - len(hardware_string))
+
+        return [hardware_string]
+
+    @tile_rpc(0x2a01, "", "H")
+    def count_tiles(self):
+        """Count the number of registered tiles including the controller"""
+        return [len(self.tile_list)]
+
+    @tile_rpc(0x2a02, "H", "3B6s6BBL")
+    def describe_tile(self, index):
+        """Describes a tile at the given index
+
+        The information returned is the exact same data as what REGISTER_TILE
+        includes as its argument.
+
+        Args:
+        - uint16_t: The index of the tile you wish to describe.  Must be less than
+            what COUNT_TILES returns.
+
+        Returns:
+        - uint8_t: a numerical hardware type identifer for the processor the tile is
+            running on.
+        - two bytes: (major, minor) API level of the tile fwutive running on this
+            tile.  The api version is encoded with the major version as a byte
+            followed by the minor version.
+        - 6-character string: The tile firmware identifier that is used to match it
+            with the correct proxy module by HardwareManager.
+        - 3 bytes: (major, minor, patch) version identifier of the application
+            firmware running on this tile.
+        - 3 bytes: (major, minor, patch) version identifier of the tile executive
+            running on this tile.
+        - uint8_t: The slot number that refers to the physical location of this tile
+            on the board.  It is used to assign the tile a fixed address on the
+            TileBus.
+        - uint32_t: A unique identifier for this tile, if the tile supports unique
+            identifiers.
+        """
+        tile = self.tiles[self.tile_list[index]]
+        return (self.config.get('hw_type', 0), self.config.get('api_major', 0), self.config.get('api_minor', 0),
+                tile.name, self.config.get('fw_major', 0), self.config.get('fw_minor', 0),
+                self.config.get('fw_patch', 0), self.config.get('exec_major', 0), self.config.get('exec_minor', 0),
+                self.config.get('exec_patch', 0), self.tile_list[index], 0)
+
+    @tile_rpc(0x1008, "", "L8xLL")
+    def controller_info(self):
+        """Get the controller UUID, app tag and os tag."""
+        return [self.config.get('iotile_id'),
+                _pack_version(*(self.config.get('os_tag', 0), self.config.get('os_version', "0.0"))),
+                _pack_version(*(self.config.get('app_tag', 0), self.config.get('app_version', "0.0")))]
+
+    @tile_rpc(0x100B, "LLBB", "L")
+    def update_os_app_tag(self, packed_os, packed_app, set_os, set_app):
+        """Update the OS and/or APP tag of the device.
```

Comment:
@mattrunchey I'm not sure I agree that we need to implement this functionality here for a couple of reasons:

- The RPC existed on PODs because there was no other way to set the os/app tag in manufacturing but for virtual devices, it's straightforward to just edit the json config file directly if needed.
- The rest of the `VirtualTile` infrastructure doesn't make any assumption that it will be reading a config from a json file or try to update it in any way so we're locking something in that preclude getting configuration from an in-memory dict or other mechanism.  

Can you explain the use-case here for why we want to add this?  I would lean toward just leaving it out and implementing the rest of the RPCs you have here absent a clear reason why we need it.

Reply:
For keeping track of node versions that was persistent, we needed some place on the file system to store the version information. Rather than implementing it outside of the virtual tile (i.e. in a task), I noticed that these RPCs existed to do that same thing for PODs. We need these RPCs somewhere, to update version information, and we decided on using a config json (if that's a bad idea, that's a separate discussion). Since version for PODs was taken care of in the controller, I figured that our VirtualDevices might benefit from being able to do the same.

So the alternative to this would be to move all of the version information to the task subsystem that we use for OTA and not add it to the core VirtualDevice side of the picture. It would be the same RPCs, just tied to the lifecycle and configuration attributes of a task compared to the controller. If other actors wanted to use it (i.e. the factory companion app), they would need to determine if the node they're talking to has an OTA subsystem, and use the RPCs from that tile instead.

I chose to put it in the virtualized controller rather than the task because it seemed that tracking the version wasn't fully in line with what the OTA updater really was responsible for.   I also assumed that the virtual_device would be more omnipresent than the OTA task, but it's possibly the case that we only really care about version when an OTA subsystem is present.  


>The rest of the VirtualTile infrastructure doesn't make any assumption that it will be reading a config from a json file or try to update it in any way so we're locking something in that preclude getting configuration from an in-memory dict or other mechanism.

I don't think this explicitly assumes there's a json configuration file -- the `self.config` always loads in these attributes from json now (with defaults if not present), but we could add a path or checker for using some in memory or `iotile registry`-like functionality alongside it. `describe_tile` as implemented has most attributes currently tied to default dummy values or config loaded files, so I would imagine something similar would have to change there too if that is the case.

Reply:
@timburke Does the above make sense? Like I said, the functionality can be moved to a particular task structure if it doesn't actually make sense to have it in the general `tile_based_device`. Let me know where you think it should be.

Reply:
I guess my gut feeling is that it should be moved to a particular task instead of in the core virtual tile infrastructure.  I think we'll end up needing to make a lot of specific tweaks in it to pull/set the os an app tag information from the correct places, which may not always be the same place.  For example, we may want to have the os tag come from a file on the root hd so it's always up to date with whatever os is installed.

Reply:
I will put it in the task then, it definitely makes more sense if we're potentially going to support a more enumerated location/source set.